WIP: Template engine concept

cernyd opened the pull request:
I noticed the a reoccuring pattern in the codebase where `Messages.template` was being called with `generate_mention(...)` to generate a response. My idea is we could instead create a template engine that will generate the mentions and fill templates automatically. This change is just a concept, I'd just like to ask if you think it is a good idea @Toaster192.

This simple "filler" utility is the first step towards a less chaotic template generation. It takes the attribute name of the template from `Messages` and automatically generates mentions with kwargs that are named `user` or `toaster`.

I currently have no way to test if the code works as I only have rubbergod set up on my other computer, so I might add some bugfixes later.

cernyd commented:
Just noticed I'm committing one of my commits from the previous PR, will fix that later.

peterdragun commented on utils.py:
```diff
@@ -34,3 +35,25 @@ def has_role(user, role_name: str):
+    if 'user' in kwargs:
+        kwargs['user'] = generate_mention(kwargs['user'])
+
+    if 'toaster' in kwargs:
```
I think using "admin" instead of "toaster" would be better.

peterdragun commented on utils.py:
```diff
@@ -34,3 +35,25 @@ def has_role(user, role_name: str):
+        template = getattr(Messages, message_name)
+        return template.format(**kwargs)
+    except Exception:
+        return ""
```
Could you specify exception that you want to catch - maybe AttributeError? Also I think returning empty string will cause error when you try to send empty message to discord (not sure about that).

Toaster192 commented on cogs/acl.py:
```diff
@@ -2,6 +2,7 @@
 from discord.ext import commands
 
 import utils
+from utils import fill_message
```
Not a fan of the 
```
import utils
from utils import fill_message
```
double import thing but the imports are kinda a clusterfuck anyways so I won't force you to change this

cernyd replied:
Not planning on leaving those there, I'll make an import revision later on. @Toaster192

cernyd commented:
I removed the double imports, we can use `utils.fill_message` for now. I also replaced all `toaster` formatting args with `admin` (sorry Toaster).  This code is not tested, testing needed.

@Toaster192 @peterdragun

Toaster192 commented on cogs/room_check.py:
```diff
@@ -13,9 +13,7 @@ def __init__(self, bot):
-            await message.channel.send(messages.bot_room_redirect.format(
-                utils.generate_mention(message.author.id),
-                config.bot_room))
+            await message.channel.send(utils.fill_message("bot_room_redirect", config.bot_room, user=message.author.id))
```
Pretty sure this won't work (`fill_message takes 1 positional argument, 2 given` or smth)

Toaster192 replied:
Also there is no user part in the formating msg so that'll have to be changed probably

peterdragun commented:
could you check if any other functions are using Messages? I mean you moved this part to utils.py, so now for example in acl.py Messages are imported but not used. This should be checked in all files. I can see at least one file(acl.py) where can this import can be removed and also one file where we still use it.

peterdragun commented on utils.py:
```diff
@@ -34,3 +35,25 @@ def has_role(user, role_name: str):
+    :return: filled template
+    """
+
+    # Convert username/toaster to a mention
```
This comment should also say "admin id" or smth like that instead of "toaster"

cernyd commented:
I reviewed the imports. I also found many more places where I did not apply the new formatting helper, that should be fixed now. Added `*args` just to be safe when using the formatter (in case some formatting args are non-positional).

peterdragun commented on features/reaction.py:
```diff
@@ -76,27 +71,17 @@ def make_embed(self, page):
-                        )
-                    )
+                    await message.channel.send(utils.fill_message("role_invalid_line",
+                                            user=message.author.id, line=line[0]))
```
can you check alignment? especially in this file, for example this line should have more indent

cernyd commented:
@peterdragun fixed indentation warnings according to the pep8 checker.
